presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1350,6 +1358,31 @@ else if (criteria instanceof JoinOn) {
             return output;
         }
 
+        private boolean needsJoinWarning(LogicalBinaryExpression expression)
+        {
+            ArrayDeque<Expression> queue = new ArrayDeque<Expression>();
+            Expression currentExpression;
+            LogicalBinaryExpression logicalExpression;
+
+            queue.add(expression);
+
+            while (!queue.isEmpty()) {
+                currentExpression = queue.remove();
+                if (currentExpression instanceof LogicalBinaryExpression) {
+                    logicalExpression = (LogicalBinaryExpression) currentExpression;
+
+                    if (logicalExpression.getOperator() == LogicalBinaryExpression.Operator.OR) {
+                        return true;
+                    }
+
+                    queue.add(logicalExpression.getLeft());
+                    queue.add(logicalExpression.getRight());
+                }
+            }
+
+            return false;
+        }
```

Comment:
I would simplify a bit by just making the parameter an `Expression`, and letting the BFS handle whether or not to traverse if the root is not a `LogicalBinaryExpression`.  And then in the above code that uses it we can remove the `instanceof` check and cast, and just use the inner if check.

```suggestion
        private boolean expressionContainsOr(Expression expression)
        {
            Deque<Expression> queue = new ArrayDeque<Expression>();
            queue.offer(expression);

            while (!queue.isEmpty()) {
                Expression currentExpression = queue.poll();
                if (currentExpression instanceof LogicalBinaryExpression) {
                    LogicalBinaryExpression logicalExpression = (LogicalBinaryExpression) currentExpression;

                    if (logicalExpression.getOperator() == LogicalBinaryExpression.Operator.OR) {
                        return true;
                    }

                    queue.offer(logicalExpression.getLeft());
                    queue.offer(logicalExpression.getRight());
                }
            }

            return false;
        }
```

Reply:
Fixed